libavcodec/libproresdecoder.c

```diff
@@ -0,0 +1,237 @@
+                memcpy(framePtr->data[0] + i * framePtr->linesize[0] + yj + 4*dataUnitBytes, &Y4, dataUnitBytes); // Y4
+                memcpy(framePtr->data[2] + i * framePtr->linesize[2] + cj + 2*dataUnitBytes, &Cr2, dataUnitBytes); // Cr2
+                memcpy(framePtr->data[0] + i * framePtr->linesize[0] + yj + 5*dataUnitBytes, &Y5, dataUnitBytes); // Y5
+            }
+        }
+    } else if (decoderctx->pix_fmt == kPRFormat_v216) {
+        // convert the v216 layout [Cb0-2B, Y0-2B, Cr0-2B, Y1-2B] to YUV422p16LE (planar YCbCr) 
+        int dataUnitBytes = 2; // read 2 bytes at a time
+        for (int i = 0; i < decoderctx->m_height; i++) {
+            for (int j = 0; j < rowBytes; j=j+dataUnitBytes) {
+                if (j % (4*dataUnitBytes) == 0) { // U component
+                    memcpy(framePtr->data[1] + i * framePtr->linesize[1] + (j / (4*dataUnitBytes))*dataUnitBytes, dataPtr + i * rowBytes + j, dataUnitBytes);
+                }
+                else if (j % (2*dataUnitBytes) == 0) { // V component.
+                    memcpy(framePtr->data[2] + i * framePtr->linesize[2] + (j / (4*dataUnitBytes))*dataUnitBytes, dataPtr + i * rowBytes + j, dataUnitBytes);
+                }
+                else if (j % dataUnitBytes == 0) { // Y component
+                    memcpy(framePtr->data[0] + i * framePtr->linesize[0] + (j / (2*dataUnitBytes))*dataUnitBytes, dataPtr + i * rowBytes + j, dataUnitBytes);
+                }
+            }
+        }
+    } else if (decoderctx->pix_fmt == kPRFormat_y416) {
+        memset(framePtr->data[3], 0, framePtr->linesize[0] * decoderctx->m_height); // A
+        // convert the y416 layout [A0 - 2B LE, Y0-2B LE, Cb0-2B LE, Cr0-2B LE] to YUVA444p16LE (planar YCbCrA) 
+        int dataUnitBytes = 2; // read 2 bytes at a time
+        for (int i = 0; i < decoderctx->m_height; i++) {
+            for (int j = 0; j < rowBytes; j=j+dataUnitBytes) {
+                int op_plane = ((j+6)%8)/2; // We want to right shift the AYCbCr order by 3 and round to get YCbCrA order
+                int write_byte = (j / (4*dataUnitBytes))*dataUnitBytes; // 4 pixels = 4*dataUnitBytes bytes and write dataUnitBytes at a time
+                memcpy(framePtr->data[op_plane] + i * framePtr->linesize[op_plane] + write_byte, dataPtr + i * rowBytes + j, dataUnitBytes);
+            }
+        }
+    }
+    free(dataPtr);
+    return res;
+};
+
+#if HAVE_THREADS
+static int update_thread_context(AVCodecContext *dstavctx, const AVCodecContext *srcavctx)
+{
+    ProResDecodeCodec *srcdecodectx = srcavctx->priv_data;
+    ProResDecodeCodec *dstdecodectx = dstavctx->priv_data;
+     
+    dstavctx->profile = srcavctx->profile;
+    dstavctx->bits_per_raw_sample = srcavctx->bits_per_raw_sample;
+    dstavctx->pix_fmt = srcavctx->pix_fmt;
+
+    dstdecodectx->m_width = srcdecodectx->m_width; 
+    dstdecodectx->m_height = srcdecodectx->m_height;
+    dstdecodectx->pix_fmt = srcdecodectx->pix_fmt;
```

Comment:
just copying over everything that I could I think of that would need to be replicated.